internal/mode/static/telemetry/collector.go

```diff
@@ -0,0 +1,156 @@
+
+// ConfigurationGetter gets the current Configuration.
+type ConfigurationGetter interface {
+	GetLatestConfiguration() *dataplane.Configuration
+}
+
+// NGFResourceCounts stores the counts of all relevant Graph resources.
+type NGFResourceCounts struct {
+	Gateways       int
+	GatewayClasses int
+	HTTPRoutes     int
+	Secrets        int
+	Services       int
+	Endpoints      int
+}
+
+// ProjectMetadata stores the name of the project and the current version.
+type ProjectMetadata struct {
+	Name    string
+	Version string
+}
+
+// Data is telemetry data.
+// Note: this type might change once https://github.com/nginxinc/nginx-gateway-fabric/issues/1318 is implemented.
+type Data struct {
+	ProjectMetadata   ProjectMetadata
+	NodeCount         int
+	NGFResourceCounts NGFResourceCounts
+}
+
+// DataCollectorConfig holds configuration parameters for DataCollectorImpl.
+type DataCollectorConfig struct {
+	// K8sClientReader is a Kubernetes API client Reader.
+	K8sClientReader client.Reader
+	// GraphGetter allows us to get the Graph.
+	GraphGetter GraphGetter
+	// ConfigurationGetter allows us to get the Configuration.
+	ConfigurationGetter ConfigurationGetter
+	// Version is the NGF version.
+	Version string
+}
+
+type DataCollectorImpl struct {
+	cfg DataCollectorConfig
+}
+
+func NewDataCollector(
+	cfg DataCollectorConfig,
+) *DataCollectorImpl {
+	return &DataCollectorImpl{
+		cfg: cfg,
+	}
+}
+
+// Collect collects and returns telemetry Data.
+func (c DataCollectorImpl) Collect(ctx context.Context) (Data, error) {
+	nodeCount, err := collectNodeCount(ctx, c.cfg.K8sClientReader)
+	if err != nil {
+		return Data{}, err
+	}
+
+	graphResourceCount, err := collectGraphResourceCount(c.cfg.GraphGetter, c.cfg.ConfigurationGetter)
+	if err != nil {
+		return Data{}, err
+	}
+
+	data := Data{
+		NodeCount:         nodeCount,
+		NGFResourceCounts: graphResourceCount,
+		ProjectMetadata: ProjectMetadata{
+			Name:    "NGF",
+			Version: c.cfg.Version,
+		},
+	}
+
+	return data, nil
+}
+
+func collectNodeCount(ctx context.Context, k8sClient client.Reader) (int, error) {
+	nodes := v1.NodeList{}
+	if err := k8sClient.List(ctx, &nodes); err != nil {
+		return 0, err
+	}
+
+	return len(nodes.Items), nil
+}
+
+func collectGraphResourceCount(
+	graphGetter GraphGetter,
+	configurationGetter ConfigurationGetter,
+) (NGFResourceCounts, error) {
+	ngfResourceCounts := NGFResourceCounts{}
+	g := graphGetter.GetLatestGraph()
+	cfg := configurationGetter.GetLatestConfiguration()
+
+	if g == nil {
+		return NGFResourceCounts{}, errors.New("latest graph cannot be nil")
+	}
+	if cfg == nil {
+		return NGFResourceCounts{}, errors.New("latest configuration cannot be nil")
+	}
+
+	if g.GatewayClass != nil {
+		ngfResourceCounts.GatewayClasses = 1
+	}
+	if g.Gateway != nil {
+		ngfResourceCounts.Gateways = 1
+	}
+	ngfResourceCounts.HTTPRoutes = len(g.Routes)
+	ngfResourceCounts.Secrets = countReferencedResources(g.ReferencedSecrets)
```

Comment:
for Secrets, even if it doesn't exist, the entry for it will still be created -- https://github.com/nginxinc/nginx-gateway-fabric/blob/cadbb72eb7244d87813da3c3682bf350ae19e8d0/internal/mode/static/state/graph/secret.go#L59-L64
so I don't think countReferencedResources will work as you expect.

At the same time, do we really need to differentiate between Referenced and Referenced by not existing resources. Since non existing is an edge case, which users will try to fix if this happens.

So I don't think we need any logic for that - I don't think we should add the function countReferencedResources and also include the whole Service in g.ReferencedServices map.

Reply:
To make sure I'm understanding correctly, you suggest that it is okay to count "referenced resources" that are non-existing? (In the case of ReferencedServices)?

Reply:
> To make sure I'm understanding correctly, you suggest that it is okay to count "referenced resources" that are non-existing? (In the case of ReferencedServices)?

yep